File: app/src/main/java/com/example/muslimmuhammad/kepoin/activity/Setup_activity.java
package com.example.muslimmuhammad.kepoin.activity;

import android.app.ProgressDialog;
import android.content.Intent;
import android.net.Uri;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;

import com.example.muslimmuhammad.kepoin.R;
import com.example.muslimmuhammad.kepoin.model.KepoPost;
import com.firebase.ui.database.FirebaseRecyclerAdapter;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.UploadTask;
import com.squareup.picasso.Callback;
import com.squareup.picasso.NetworkPolicy;
import com.squareup.picasso.Picasso;
import com.theartofdev.edmodo.cropper.CropImage;
import com.theartofdev.edmodo.cropper.CropImageView;

public class Setup_activity extends AppCompatActivity {
    private ImageButton mSetupImageBtn;
    private EditText mNameField;
    private Button mSubmitBtn;
    private Uri mImageUri = null;
    private DatabaseReference mDatabaseUser;
    private static final int GALLERY_REQUEST = 1;
    private FirebaseAuth mAuth;
    private StorageReference mStorageImage ;
    private ProgressDialog mProgress;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_setup_activity);
        mAuth = FirebaseAuth.getInstance();
        mStorageImage = FirebaseStorage.getInstance().getReference().child("Profil_image");
        mDatabaseUser = FirebaseDatabase.getInstance().getReference().child("Users");
        mSubmitBtn = (Button)findViewById(R.id.btn_input);
        mNameField = (EditText) findViewById(R.id.edt1);
        mSetupImageBtn = (ImageButton)findViewById(R.id.imageBro);
        mProgress = new ProgressDialog(this);

        mSetupImageBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent galleryIntent = new Intent();
                galleryIntent.setAction(Intent.ACTION_GET_CONTENT);
                galleryIntent.setType("image/*");
                startActivityForResult(galleryIntent,GALLERY_REQUEST);

            }
        });
        mSubmitBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startSetupAccount();
            }
        });


    }

    private void startSetupAccount() {

        final String name = mNameField.getText().toString().trim();
        final String user_id = mAuth.getCurrentUser().getUid();


        if(!TextUtils.isEmpty(name) && mImageUri != null){
            mProgress.setMessage("Finishing setup...");
            mProgress.show();
            StorageReference filePath = mStorageImage.child(mImageUri.getLastPathSegment());
            filePath.putFile(mImageUri).addOnSuccessListener(new OnSuccessListener<UploadTask.TaskSnapshot>() {
                @Override
                public void onSuccess(UploadTask.TaskSnapshot taskSnapshot)
                {
                    String downloadUri = taskSnapshot.getDownloadUrl().toString();
                    mDatabaseUser.child(user_id).child("name").setValue(name);
                    mDatabaseUser.child(user_id).child("image").setValue(downloadUri);
                    mProgress.dismiss();
                    Intent mainIntent = new Intent(Setup_activity.this,KepoIn.class);
                    mainIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
                    startActivity(mainIntent);
                }
            });


        }

    }

    @Override
    protected void onStart() {
        super.onStart();
        loadphoto();

    }
    private void loadphoto(){
        mDatabaseUser.child(mAuth.getCurrentUser().getUid()).addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {
                String username = (String) dataSnapshot.child("name").getValue();

                String post_image = (String) dataSnapshot.child("image").getValue();


                mNameField.setText(username);

                //Picasso.with(Setup_activity.this).load(post_image).placeholder(R.drawable.common_full_open_on_phone).into(mSetupImageBtn);

//                Picasso.with(Setup_activity.this).load(post_image).into(mSetupImageBtn);
//                mSetupImageBtn.setImageResource();

            }

            @Override
            public void onCancelled(DatabaseError databaseError) {

            }
        });



    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if(requestCode == GALLERY_REQUEST && resultCode == RESULT_OK ){
            mImageUri = data.getData();
            Uri imageUri = data.getData();
            CropImage.activity(imageUri)
                    .setActivityTitle("Crop your image")
                    .setGuidelines(CropImageView.Guidelines.ON)
                    .setAspectRatio(1,1)
                    .setMaxCropResultSize(1000,1000)
                    .setMinCropResultSize(100,100)
                    .start(this);
        }
        if (requestCode == CropImage.CROP_IMAGE_ACTIVITY_REQUEST_CODE) {
            CropImage.ActivityResult result = CropImage.getActivityResult(data);
            if (resultCode == RESULT_OK) {
                mImageUri = result.getUri();
                mSetupImageBtn.setImageURI(mImageUri);
            } else if (resultCode == CropImage.CROP_IMAGE_ACTIVITY_RESULT_ERROR_CODE) {
                Exception error = result.getError();
            }
        }
    }
}

File: app/src/main/java/com/achmad/sun3toline/kepoin/activity/SetupActivity.kt
/*
 * *
 *  * Created by Achmad Fathullah on 10/3/20 9:21 PM
 *  * Copyright (c) 2020 . All rights reserved.
 *  * Last modified 10/3/20 8:20 PM
 *
 */

package com.achmad.sun3toline.kepoin.activity

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.widget.EditText
import android.widget.ImageButton
import androidx.appcompat.app.AppCompatActivity
import com.achmad.sun3toline.kepoin.R
import com.achmad.sun3toline.kepoin.utils.dialog.ProgressDialog
import com.achmad.sun3toline.kepoin.utils.extension.showToast
import com.bumptech.glide.Glide
import com.google.android.material.button.MaterialButton
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageReference
import com.skydoves.whatif.whatIfNotNull
import com.theartofdev.edmodo.cropper.CropImage
import com.theartofdev.edmodo.cropper.CropImageView

class SetupActivity : AppCompatActivity() {
    private var mSetupImageBtn: ImageButton? = null
    private var mNameField: EditText? = null
    private var mSubmitBtn: MaterialButton? = null
    private var mImageUri: Uri? = null
    private var mDatabaseUser: DatabaseReference? = null
    private var mAuth: FirebaseAuth? = null
    private var mStorageImage: StorageReference? = null
    private var mProgress: ProgressDialog? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_setup_activity)
        mAuth = FirebaseAuth.getInstance()
        mStorageImage = FirebaseStorage.getInstance().reference.child("Profil_image")
        mDatabaseUser = FirebaseDatabase.getInstance().reference.child("Users")
        mSubmitBtn = findViewById(R.id.btn_input)
        mNameField = findViewById(R.id.edt1)
        mSetupImageBtn = findViewById(R.id.imageBro)
        mProgress = ProgressDialog(this)
        mSetupImageBtn?.setOnClickListener {
            val galleryIntent = Intent()
            galleryIntent.action = Intent.ACTION_GET_CONTENT
            galleryIntent.type = "image/*"
            startActivityForResult(galleryIntent, GALLERY_REQUEST)
        }
        mSubmitBtn?.setOnClickListener { startSetupAccount() }
    }

    private fun startSetupAccount() {
        val name = mNameField?.text.toString().trim { it <= ' ' }
        val userId = mAuth?.currentUser?.uid.toString()
        mImageUri.whatIfNotNull {
            mProgress?.setMessage("Finishing setup...")
            mProgress?.show()
            val filePath = mStorageImage?.child(it.lastPathSegment.toString())
            filePath?.putFile(it)?.addOnSuccessListener {
                filePath.downloadUrl.addOnSuccessListener { taskSnapshot ->
                    val downloadUri: Uri = taskSnapshot
                    mDatabaseUser?.child(userId)?.child("name")?.setValue(name)
                    mDatabaseUser?.child(userId)?.child("image")
                            ?.setValue(downloadUri)
                    mProgress?.dismiss()
                    val mainIntent = Intent(this@SetupActivity, KepoIn::class.java)
                    mainIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
                    startActivity(mainIntent)
                }

            }
        }
    }

    override fun onStart() {
        super.onStart()
        loadPhoto()
    }

    private fun loadPhoto() {
        mDatabaseUser?.child(mAuth?.currentUser?.uid.toString())?.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                val username = dataSnapshot.child("name").value as String?
                val postImage = dataSnapshot.child("image").value as String?
                mNameField?.setText(username)
                mSetupImageBtn?.let { Glide.with(this@SetupActivity).load(postImage).placeholder(R.drawable.common_full_open_on_phone).into(it) }
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == GALLERY_REQUEST && resultCode == RESULT_OK) {
            data?.data.whatIfNotNull(
                    whatIf = {
                        mImageUri = it
                        CropImage.activity(it)
                                .setActivityTitle("Crop your image")
                                .setGuidelines(CropImageView.Guidelines.ON)
                                .setAspectRatio(1, 1)
                                .setMaxCropResultSize(1000, 1000)
                                .setMinCropResultSize(100, 100)
                                .start(this)
                    },
                    whatIfNot = {
                        showToast(R.string.error_take_image)
                    }
            )
        }
        if (requestCode == CropImage.CROP_IMAGE_ACTIVITY_REQUEST_CODE) {
            val result = CropImage.getActivityResult(data)
            if (resultCode == RESULT_OK) {
                mImageUri = result.uri
                mSetupImageBtn?.setImageURI(mImageUri)
            } else if (resultCode == CropImage.CROP_IMAGE_ACTIVITY_RESULT_ERROR_CODE) {
                showToast(result.error.toString())
            }
        }
    }

    companion object {
        private const val GALLERY_REQUEST = 1
    }
}